presto-main/src/main/java/com/facebook/presto/operator/MergeOperator.java

```diff
@@ -0,0 +1,338 @@
+        this.sourceId = requireNonNull(sourceId, "sourceId is null");
+        this.exchangeClientSupplier = requireNonNull(exchangeClientSupplier, "exchangeClientSupplier is null");
+        this.pagesSerde = requireNonNull(pagesSerde, "pagesSerde is null");
+        this.comparator = requireNonNull(comparator, "comparator is null");
+        this.outputChannels = requireNonNull(outputChannels, "outputChannels is null");
+        this.outputTypes = requireNonNull(outputTypes, "outputTypes is null");
+        this.pageBuilder = new PageBuilder(outputTypes);
+    }
+
+    @Override
+    public PlanNodeId getSourceId()
+    {
+        return sourceId;
+    }
+
+    @Override
+    public Supplier<Optional<UpdatablePageSource>> addSplit(Split split)
+    {
+        requireNonNull(split, "split is null");
+        checkArgument(split.getConnectorId().equals(REMOTE_CONNECTOR_ID), "split is not a remote split");
+
+        URI location = ((RemoteSplit) split.getConnectorSplit()).getLocation();
+        locations.add(location);
+        return Optional::empty;
+    }
+
+    @Override
+    public void noMoreSplits()
+    {
+        ImmutableList.Builder<PageSupplier> pageSuppliers = ImmutableList.builder();
+        for (URI location : locations) {
+            ExchangeClient exchangeClient = closer.register(exchangeClientSupplier.get());
+            exchangeClient.addLocation(location);
+            exchangeClient.noMoreLocations();
+            pageSuppliers.add(new RemoteExchangePageSupplier(exchangeClient, pagesSerde));
+        }
+
+        mergeSortProcessor = new MergeSortProcessor(comparator, pageSuppliers.build());
+        blockedOnSplits.set(null);
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public List<Type> getTypes()
+    {
+        return outputTypes;
+    }
+
+    @Override
+    public void finish()
+    {
+        close();
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return closed || (mergeSortProcessor != null && mergeSortProcessor.isFinished() && pageBuilder.isEmpty());
+    }
+
+    @Override
+    public ListenableFuture<?> isBlocked()
+    {
+        if (!blockedOnSplits.isDone()) {
+            return blockedOnSplits;
+        }
+        return mergeSortProcessor.isBlocked();
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return false;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        throw new UnsupportedOperationException(getClass().getName() + " can not take input");
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        checkState(mergeSortProcessor != null, "mergeSources must be non-null when getting input");
+
+        if (closed) {
+            return null;
+        }
+
+        while (!pageBuilder.isFull()) {
+            PageWithPosition pageWithPosition = mergeSortProcessor.poolNextRow();
```

Comment:
Yeah... I can't think of anything different without making PageWithPosition more complicated, and I can see why you would want it to be simple. We can leave it.